### PR TITLE
merge --all

### PR DESCRIPTION
**Pull request by v1kko**
This PR introduces a `merge -all` command, and simplifies the existing `merge` command

- the `cfg.merge.steps` config variable is removed in favor of `cfg.merge.variables`. duqtools can now figure out all the dimensions and stuff if necessary

- [ ] Optional, we still have to find a way to check the quality of the standard distribution, but we could add that later to duqtools if requested

### CODE REVIEW

**Review by stefsmeets — CHANGES_REQUESTED** [review]
Nice work, took me a while to review this PR, it does a lot 😅

I left some comments in the code where I felt that it was not entirely clear to me what is happening.

This PR will break the dashboard. Because the api for `merge_data` changes, the dashboard should be updated too.

Just a thought: `duqtools merge` enters new terrain by mixing values from `duqtools.yaml` with cli parameters. `-all` overrides what is defined in the config. I'm just wondering if this mixed interface is something that we want, or is this confusing? Should we allow variables to merge to be defined on the CLI as well?

**Comment by v1kko** [comment]
> Nice work, took me a while to review this PR, it does a lot sweat_smile
> 
> I left some comments in the code where I felt that it was not entirely clear to me what is happening.
> 
> This PR will break the dashboard. Because the api for `merge_data` changes, the dashboard should be updated too.
> 
> Just a thought: `duqtools merge` enters new terrain by mixing values from `duqtools.yaml` with cli parameters. `-all` overrides what is defined in the config. I'm just wondering if this mixed interface is something that we want, or is this confusing? Should we allow variables to merge to be defined on the CLI as well?

Yes, actually it would make sense to define `merge` more like the `plot` command that we already have. And we could remove the `merge:` section from the config entirely. In that way we unify the API for merge (also for the dashboard) again, and there are less surprises

File: src/duqtools/ids/_merge.py [hunk 1]
```diff
@@ -1,83 +1,84 @@
...
+    variable_dict = {variable.name: variable for variable in variables}
+    for variable in variables:
+        for dim in variable.dims:
+            # TODO time should also be $ prefixed, remove the `or time` when fixed
+            if not (dim.startswith('$') or dim.startswith('time')):
+                continue
+            name = dim.lstrip('$')
+            if name not in variable_dict.keys():
+                variables.append(var_lookup[name])
+                variable_dict[name] = variables[-1]
```
[the remark this answers is not in the document]
Comment (v1kko) [reply src/duqtools/ids/_merge.py:48]: 1. its not _that_ unreadable, but that's subjective :smiley:
2. I agree, I will see if it can be hidden elsewhere

**Comment by stefsmeets** [comment]
> Yes, actually it would make sense to define `merge` more like the `plot` command that we already have. And we could remove the `merge:` section from the config entirely. In that way we unify the API for merge (also for the dashboard) again, and there are less surprises

I think it's also helpful with #432 to do merging for large scale validation.

File: src/duqtools/ids/_merge.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (stefsmeets) [reply src/duqtools/ids/_merge.py:48]: re: 2

> Perhaps we should even have a var_lookup class that can deal with these cases and transparently pick up $rho_tor_norm and rho_tor_norm as the same variable.

This ties in with what I'm doing for #418 already, so I can take care of that part.

File: src/duqtools/ids/_mapping.py [hunk 2]
```diff
@@ -360,9 +362,44 @@ def _contains_empty(arr):
...
+            self[path] = data
+            return
+
+        root, sub, *remaining = parts
+        nodes = self[root]
+        for index in range(len(nodes)):
+            path = f'{root}/{index}/{sub}'
+            self._write_back(data[index], path, *remaining)
+
+    def write_back(self, variable: str, data: xr.DataArray) -> None:
```
[the remark this answers is not in the document]
Comment (v1kko) [reply src/duqtools/ids/_mapping.py:388]: You are correct that it is the inverse of _fill_array_from_parts, but its only 7 lines of code, and sufficiently distinct (and hard to merge a fetch and writeback operation) to leave it two separate functions.

File: src/duqtools/merge.py [hunk 3]
```diff
@@ -2,29 +2,31 @@
...
 
-        queue_merge_data(source_data=handles,
-                         target=target,
-                         time_var=step.time_variable,
-                         grid_var=step.grid_variable,
-                         data_vars=step.data_variables)
+    if kwargs['all']:
+        _, variables = partition(lambda var: var.type == 'IDS-variable',
+                                 var_lookup.values())
+        variables = list(variables)
```
[the remark this answers is not in the document]
Comment (v1kko) [reply src/duqtools/merge.py:34]: yes

File: src/duqtools/ids/_merge.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (v1kko) [reply src/duqtools/ids/_merge.py:48]: I will leave this until #418 is ready

File: src/duqtools/ids/_merge.py [hunk 4]
```diff
@@ -1,83 +1,84 @@
...
+
+        # Now we have to get the stddeviations
+        mean_data = ids_data.mean(dim='handle')
+        std_data = ids_data.std(dim='handle')
+
+        # Then, write it back to target
+        for name in ids_data.data_vars.keys():
+            target_ids.write_back(variable_dict[name].path, mean_data[name])
+            target_ids.write_back(variable_dict[name].path + '_error_upper',
+                                  std_data[name])
```
[the remark this answers is not in the document]
Comment (v1kko) [reply src/duqtools/ids/_merge.py:83]: Correct, well, coincidentally this is how the data is stored in HDF5, and would be much faster, but I don't think that the IMAS Access Layer supports this type of variable access yet, unfortunately. If it becomes a bottleneck, we can look at it again

File: src/duqtools/ids/_merge.py [hunk 5]
```diff
@@ -1,83 +1,84 @@
...
+                variables.append(var_lookup[name])
+                variable_dict[name] = variables[-1]
+
+    # Get all known variables per ids
+    grouped_ids_vars = groupby(variables, keyfunc=lambda var: var.ids)
+
+    for ids_name, variables in grouped_ids_vars.items():
+
+        # Get all data, and rebase it
+        idss = [handle.get(ids_name) for handle in handles]  # type: ignore
```
[the remark this answers is not in the document]
Comment (v1kko) [reply src/duqtools/ids/_merge.py:56]: Yes good one

**Comment by v1kko** [comment]
@stefsmeets I think everything that can/should be done later is in issues now, so lets merge this?

**Pull request merged**